### PR TITLE
Output logs in Delayed Job's log file

### PR DESCRIPTION
**Pull request by sauloperez**
#### What? Why?

Related to #5294

There's a bunch of logging bits we need to find out what happened in #5294, the first of which is logging as we are used to for Rails itself. Turns out we were already logging but the log lines weren't displayed.

That's because workers have no access to the Rails logger unless we hack things a bit in the `config/initializer`. If we want to keep DJ's logs in a different file, which I think is good for debugging purposes, we need to diverge from `Rails.logger`.

However, there's no access to `Delayed::Worker.logger` in custom job classes either so there's no way other than mimicking that logger's format if we want log lines to be consistent.

Now the current logs will see the light like:

```
2020-05-07T14:09:42+0000: [Worker(host:8dc5e7e6f434 pid:1)] Job SubscriptionPlacementJob (id=54) RUNNING
2020-05-07T14:09:42+0000: Placing Order for Proxy Order 1
2020-05-07T14:09:42+0000: Issue in Subscription Order 2: complete
2020-05-07T14:09:42+0000: Placing Order for Proxy Order 7
2020-05-07T14:09:42+0000: Issue in Subscription Order 3: complete
2020-05-07T14:09:43+0000: [Worker(host:8dc5e7e6f434 pid:1)] Job SubscriptionPlacementJob (id=54) COMPLETED after 0.3378
2020-05-07T14:09:43+0000: [Worker(host:8dc5e7e6f434 pid:1)] 1 jobs processed at 2.8054 j/s, 0 failed
```

We'll see if we need to format these messages differently to make them more greppeable once we use them.

#### What should we test?

We need to check output is logged when running `SubscriptionConfirmJob` and `SubscriptionPlacementJob`. The easiest way is to set up an OC a schedule and a subscription and then:

* Trigger subscription placement by running `Delayed::Job.enqueue SubscriptionPlacementJob.new` from a Rails console
* Trigger subs confirmation by running `Delayed::Job.enqueue SubscriptionConfirmJob.new` from a Rails console

All log lines should prefixed with a timestamp in a consistent format.

#### Release notes

Output log messages in `log/delayed_job.rb` for all susbcription's background jobs

Changelog Category: Fixed

### CODE REVIEW

File: app/jobs/subscription_placement_job.rb [hunk 1]
```diff
@@ -87,4 +87,8 @@ def send_empty_email(order, changes)
     record_issue(:empty, order)
     SubscriptionMailer.empty_email(order, changes).deliver
   end
+
+  def logger
+    Delayed::Worker.logger
```
Comment (sauloperez) [inline app/jobs/subscription_placement_job.rb:92]: it turns out we can't access the logger from a custom job :cry: :point_right: https://github.com/collectiveidea/delayed_job/blob/b78f88f11ed09b3c952406b08c2cc57e8e5b63bb/lib/delayed/backend/base.rb#L77-L90, so we can't use their formatting either :sob: 

What I'm considering is going with https://api.rubyonrails.org/classes/ActiveSupport/TaggedLogging.html in 
https://github.com/openfoodfoundation/openfoodnetwork/blob/74c7a01151d527e9f1f09fc5859e4faf66cb5a1e/config/initializers/delayed_job.rb#L1

so `Delayed::Worker.logger` gets much reacher formatting capabilities we can use ourselves.

Comment (luisramos0) [reply app/jobs/subscription_placement_job.rb:92]: yeah, that sounds good to me!

Comment (luisramos0) [reply app/jobs/subscription_placement_job.rb:92]: @mkllnk used tagged logging in the sample data:
https://github.com/openfoodfoundation/openfoodnetwork/blob/master/lib/tasks/sample_data/logging.rb

**Comment by mkllnk** [comment]
In Fair Food we have this simple line in `config/initializers/delayed_job.rb`:
```rb
Delayed::Worker.logger = Logger.new(File.join(Rails.root, "log", "delayed_job.log"))
```

It works really well.

**Comment by sauloperez** [comment]
yes, @mkllnk we do too

https://github.com/openfoodfoundation/openfoodnetwork/blob/c2a9a698f494c3c1accb054029a692ffcbb1b3ce/config/initializers/delayed_job.rb#L1

 but worker.logger != job.logger. That's the underlying problem here.

File: app/jobs/job_logger.rb [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+                   logger = Delayed::Worker.logger.clone
+                   logger.formatter = Formatter.new
+                   logger
+                 end
+  end
+
+  def self.info(name, message)
+    logger.info(name) { "#{message}" }
+  end
+end
```
Comment (sauloperez) [inline app/jobs/job_logger.rb:20]: With this, things look like:

```
2020-05-06T15:42:24+0000: [Worker(host:0ee4c04a3c6e pid:1)] Starting job worker
2020-05-06T15:42:24+0000: [Worker(host:0ee4c04a3c6e pid:1)] Job SubscriptionPlacementJob (id=50) RUNNING
2020-05-06T15:42:24+0000: [SubscriptionPlacementJob] Placing Order for Proxy Order 1
2020-05-06T15:42:24+0000: [] Issue in Subscription Order 2: complete
2020-05-06T15:42:24+0000: [SubscriptionPlacementJob] Placing Order for Proxy Order 7
2020-05-06T15:42:24+0000: [] Issue in Subscription Order 3: complete
2020-05-06T15:42:25+0000: [Worker(host:0ee4c04a3c6e pid:1)] Job SubscriptionPlacementJob (id=50) COMPLETED after 1.2083
2020-05-06T15:42:25+0000: [Worker(host:0ee4c04a3c6e pid:1)] 1 jobs processed at 0.8113 j/s, 0 failed
```

**Comment by luisramos0** [comment]
It's looking good! Let me know when this is ready for review.

**Comment by sauloperez** [comment]
Done! I decided to drop the tagged logging to show the job class name because the code grew too much in complexity and it didn't seem worth it since the worker already shows the job name.

**Comment by luisramos0** [comment]
you want to move it to Code Review right?

File: engines/order_management/app/services/order_management/subscriptions/proxy_order_syncer.rb [hunk 3]
```diff
@@ -38,14 +38,14 @@ def sync_subscriptions!
 
       def initialise_proxy_orders!
         uninitialised_order_cycle_ids.each do |order_cycle_id|
-          Rails.logger.info "Initializing Proxy Order " \
+          JobLogger.logger.info "Initializing Proxy Order " \
```
Comment (luisramos0) [inline engines/order_management/app/services/order_management/subscriptions/proxy_order_syncer.rb:41]: This is not called from a Job, so this should be a normal logger call.

File: spec/jobs/subscription_placement_job_spec.rb [hunk 4]
```diff
@@ -54,9 +54,15 @@
       end
 
       it "processes placeable proxy_orders" do
+        expect(JobLogger.logger).to receive(:info)
         job.perform
         expect(job).to have_received(:place_order).with(proxy_order.reload.order)
       end
+
+      it "logs using JobLogger" do
```
Comment (luisramos0) [inline spec/jobs/subscription_placement_job_spec.rb:62]: I dont see the point of testing the logger calls. but I am fine with it if you think it's worth the extra lines of code :+1:

Comment (sauloperez) [reply spec/jobs/subscription_placement_job_spec.rb:62]: well, me neither. It's something I don't usually test but I needed a way to quickly develop and not boot a worker every time. I can remove them if you prefer. It already served its purposes.

Comment (luisramos0) [reply spec/jobs/subscription_placement_job_spec.rb:62]: ah, I see :+1: it's fine. we keep them.

Comment (sauloperez) [reply spec/jobs/subscription_placement_job_spec.rb]: solved

**Comment by sauloperez** [comment]
I'm not seeing any custom log lines in staging after creating a subscription and having the OC closed. Even after forcing the following from a console

```
ProxyOrder.update_all(placed_at: nil)
Delayed::Job.enqueue SubscriptionPlacementJob.new
```

In both cases I get

```
2020-05-08T11:48:08+0200: [Worker(delayed_job host:vps189799 pid:21080)] Job SubscriptionPlacementJob (id=214225) RUNNING
2020-05-08T11:48:08+0200: [Worker(delayed_job host:vps189799 pid:21080)] Job SubscriptionPlacementJob (id=214225) COMPLETED after 0.0564
2020-05-08T11:48:08+0200: [Worker(delayed_job host:vps189799 pid:21080)] 1 jobs processed at 8.4009 j/s, 0 failed
2020-05-08T11:50:08+0200: [Worker(delayed_job host:vps189799 pid:21080)] Job SubscriptionPlacementJob (id=214226) RUNNING
2020-05-08T11:50:08+0200: [Worker(delayed_job host:vps189799 pid:21080)] Job SubscriptionPlacementJob (id=214226) COMPLETED after 0.0223
2020-05-08T11:50:08+0200: [Worker(delayed_job host:vps189799 pid:21080)] Job SubscriptionConfirmJob (id=214227) RUNNING
2020-05-08T11:50:08+0200: [Worker(delayed_job host:vps189799 pid:21080)] Job SubscriptionConfirmJob (id=214227) COMPLETED after 0.0290
```

So I'll need to investigate this further considering staging's logging settings. Might be the log level?

**Comment by luisramos0** [comment]
were the orders created? maybe just a problem with making subs work?
https://github.com/openfoodfoundation/openfoodnetwork/wiki/Testing-Subscriptions

**Comment by sauloperez** [comment]
It did work:

```
2020-05-12T09:08:48+0200: Placing Order for Proxy Order 18
2020-05-12T09:08:49+0200: Issue in Subscription Order 1024: empty
```
What I noticed is that these lines get logged in `log/delayed_job.log` regardless of how they are executed from DJ or not. In production that happens from DJ only but that's something to be aware of.

**Comment by luisramos0** [comment]
ah, I was going through this and I think you maybe have had the same problem I did: I didnt read the log message :-)
"**Issue in Subscription** Order 1025: empty "
That's why there was no confirmation job logs.

The product was out of stock.
I added stock and run the processes again (the wiki page is really useful!) and voilá:
`2020-05-12T11:03:35+0200: Confirming Order for Proxy Order 18`

I am merging :tada:

**Comment by sauloperez** [comment]
> the wiki page is really useful!

indeed! thanks for creating it!

**Pull request merged**